contrib: review dev env and demo project setup (#1928)
iterative/vscode-dvc: CONTRIBUTING.md

## Development environment

Setting up a dev environment allows contributors to test changes to this
extension. The source code is built, and the resulting package loaded into a
special instance of VS Code.

First, ensure that [Visual Studio Code](https://code.visualstudio.com) and

> [Yarn](https://yarnpkg.com/) are installed.

Reviewer: Mention `yarn install` dependencies?

Reviewer: trying to be as explicit as possible

Author: Added `yarn install` (from root).

- Open this repository as a project in VS Code.

- Run `Tasks: Run Build Task` (Shift + Ctrl/Cmd + `b`) to start the extension
  and Webview dev servers (alternatively, run `yarn dev-server` from the
  _Integrated Terminal_).

  > **Warning**: Having a separate `.vsix` version of this extension installed

>   > may cause all kinds of chaos in your development environment.

Author: Hey @mattseddon 👋🏼 what is this warning meant to prevent? `Tasks: Run Build Task` doesn't seem to generate a .vsix file.

Author: Also, I think it's just `yarn dev` now?

Reviewer: definitely dev-server: `"dev-server": "yarn turbo run dev --parallel",`

Reviewer: If you have the extension installed (either through a `VSIX` or the marketplace) and start a debug session with the `dev-server` running you're going to be very confused as to what is going on.

Author: I have the extension installed via marketplace.

Shift + Cmd + `b` seems to use `yarn run dev`:

<img width="1072" alt="image" src="https://user-images.githubusercontent.com/1477535/176597587-8d355e80-b04c-47a9-a7fc-9a4679fd0bc6.png">

Ends with

```
webpack 5.73.0 compiled successfully in 5388 ms
No issues found.
```

If I instead run yarn `dev-server` from Terminal, the output looks a bit different but the ending is similar:

```
dvc-vscode-webview:dev: webpack 5.73.0 compiled successfully in 5754 ms
dvc-vscode-webview:dev: No issues found.
```

Start Debugging in either case seems to work fine. The dev host disables all installed extensions anyway:

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/1477535/176597806-7201eb62-4de5-4756-8915-090f91a37a79.png">

Then again the same happens if I don't run any build task before F5 😅 I've no idea what's happening anyway.

Bottom line: are these instructions clear for extension developers?

Author: Then again the same happens if I don't run any build task before F5 😅 I've no idea what's happening anyway.

Bottom line: are these instructions clear for extension developers?

Author: > the same happens if I don't run any build task before F5

Ah, after deleting extension/dist F5 fails.

Reviewer: > Start Debugging after building (with or without webview's webpack watch) seems to work fine. The dev host disables all installed extensions anyway:

This ties in with https://github.com/iterative/vscode-dvc/pull/1928#discussion_r910635641

  > DVC project uses an isolated Python env (see [this warning](#warning)).

- Open a DVC project in the _Extension Development Host_. VS Code will remember
  the last project opened, so this step only has to be done once.

  > **Note**: We have provided a demo project as part of this repo, but feel
  > free to use any DVC project that you have available.

[extension development host]:
  https://code.visualstudio.com/api/working-with-extensions/testing-extension

## The demo project

The [demo project](demo) is provided as a lightweight, convenient testbed to try
your changes to this extension.

> **Note**: It is not an exhaustive showcase of DVC's features. Testers are
> encouraged to try other DVC repositories -- especially real-world cases!

- Run `yarn setup:venv` from the _Terminal_ in the root of this repo to set up a
  Python virtual environment for the demo project (in `demo/.env`).

- Open the `./demo` project in a VS Code window, for example in the _Extension
  Development Host_ (see previous section).

- Pull the project data using this extension (either from the **DVC Tracked**
  panel in the _File Explorer_ or the **DVC panel** in _Source Control_) or by

>   running `dvc pull` from _Terminal_.

Reviewer: from the _Integrated Terminal_?

Author: https://code.visualstudio.com/docs/editor/integrated-terminal

> If you use a vs code terminal at this point, we know you're in the project root.

- In order to [run experiments] in the demo project, the [Python extension]
  should be installed so that the virtual env is activated.

[python extension]:
  https://marketplace.visualstudio.com/items?itemName=ms-python.python
[run experiments]:
  https://github.com/iterative/vscode-dvc/blob/main/extension/resources/walkthrough/run-experiments.md

<a id='warning'></a>

> **Warning**: When using any project that relies on an isolated Python
> environment (`conda`, `venv`, etc.), Microsoft's Python extension is required.

> > It's used by this extension to locate and utilize the required environment.

Author: I feel like this note should be moved out to the README or some other place since it can apply to any DVC project and not just to the demo/ project or contrib guidelines. Right?

Reviewer: Agreed, I would include it in the Development environment section or at least closer to the top of this one. I think it is ok to duplicate

## Development environment

Setting up a dev environment allows contributors to test changes to this
extension. The source code is built, and the resulting package loaded into a
special instance of VS Code.

First, ensure that [Visual Studio Code](https://code.visualstudio.com) and
[Yarn](https://yarnpkg.com/) are installed.

- Open this repository as a project in VS Code and run `yarn install` from the
  _Integrated Terminal_.

- Run `Tasks: Run Build Task` (Shift + Ctrl/Cmd + `b`) to start the extension
  and Webview dev servers (alternatively, run `yarn dev-server` from the
  _Terminal_). The latest build is saved in `extension/dist`.

  > **Warning**: Having a separate `.vsix` version of this extension installed
  > may cause all kinds of chaos in your dev env.

- **Start Debugging** (F5) to open the [Extension Development Host], a child
  instance of VS Code with the local extension build installed.

  > **Note**: using the `Run Extension` command when running the debugger will
  > prevent all other extensions from loading into VS Code. This will improve
  > the performance of VS Code but can cause certain DVC commands to fail if the

>   > DVC project uses an isolated Python env (see [this warning](#warning)).

Author: Actually, this warning I'm also not getting: What is the `Run Extension` command?

<img width="955" alt="image" src="https://user-images.githubusercontent.com/1477535/176603417-bfd9f75a-6149-48b9-987c-633be6e4aa9e.png">

Reviewer: The options are in a different place for the debugger: 
https://user-images.githubusercontent.com/37993418/176607460-3c1be14f-d168-4651-aa9f-c0045cf455d2.mov

## The demo project

The [demo project](demo) is provided as a lightweight, convenient testbed to try
the extension out with. It is not an exhaustive showcase of DVC's features, so
testers are encouraged to try other DVC repositories -- especially real-world
cases!

> - Install [Yarn](https://yarnpkg.com/) if needed. Run `yarn setup:venv` from the

Reviewer: [I] If you are going to go as far as saying install yarn you also need to mention run `yarn install` or `yarn setup:venv` will not work.

Reviewer: We already state to install yarn on line 22

Author: Removing "Install yarn" to simplify.

> Personally I'd replace this whole bullet with a more generic instruction but don't want to deal with the implications rn. See https://github.com/iterative/vscode-dvc/pull/1928#discussion_r903058889 above

  root of this repo to set up a virtual Python environment for the demo project
  (in `demo/.env`).

- Important: Open the `demo/` folder of this repo directly in VS Code.

- Pull the project data with the extension (either from the **DVC Tracked**
  panel in the _File Explorer_ or the **DVC panel** in the _Source Code_ view)
  or by running `dvc pull` from a VS Code _Terminal_.

- In order to [run experiments] in the demo project, the virtual env should be
  loaded through the Microsoft [Python extension] (please install it), having
  this in `.vscode/settings.json`:

  ```json

>   "python.defaultInterpreterPath": "demo/.env/bin/dvc"

Reviewer: [I] This is the option for manually specifying a [Python interpreter](https://code.visualstudio.com/docs/python/environments#_manually-specify-an-interpreter) for the Python extension. Why would you point it at the demo project's version of DVC?

Author: Typo, should be `python`. Fixing

## The demo project

The [demo project](demo) is provided as a lightweight, convenient testbed to try
the extension out with. It is not an exhaustive showcase of DVC's features, so
testers are encouraged to try other DVC repositories -- especially real-world
cases!

- Install [Yarn](https://yarnpkg.com/) if needed. Run `yarn setup:venv` from the
  root of this repo to set up a virtual Python environment for the demo project
  (in `demo/.env`).

> - Important: Open the `demo/` folder of this repo directly in VS Code.

Reviewer 2: hey @iterative/vs-code - do you primarily open it directly when you run the dev version or from the extension root?

Reviewer: The normal process is to 

1. Open the project's root
2. Start the dev server
3. Hit f5 to start a debug session
4. Open the demo project in the debug session

Author: Thanks. From this I get that you tend to open the demo/ project directly when using it to test changes to the extension (which is `demo`'s main purpose).

> BTW, I'm having [this issue again](https://github.com/iterative/dvc-checkpoints-mnist/issues/20) in the demo/ project when the branch is not `main` -- I assume contributors will be in new branches when doing all this. I'll try to fix it and if not reach out in Slack for now.

>   a Python virtual environment in the demo project (in `demo/.env`).

Author: Does it work if you `yarn setup:venv` first from the root, and then `cd` into `demo/` for everything else? Instead of going back and forth in each step... ⏳

- Pull the project data with the extension (either from the **DVC Tracked**
  panel in the _File Explorer_ or the **DVC panel** in the _Source Code_ view)
  or by running `dvc pull` from a VS Code _Terminal_.

>   > Remember to activate the virtual env first (`source .env/bin/activate`) if

Reviewer 2: it should be activated automatically if you run Terminal in VS Code

Author: "if you're using your system's terminal" 🙂 

Is the note confusing/unnecessary ? I don't mind removing it.

Author: Clarified further for now ("... system terminal outside of the IDE")

Reviewer 2: yes, seems premature and level of details doesn't correlate with the other places in the document

Author: Removed.

  > you're using your system's terminal.

> - In order to run experiments in the demo project (from the **DVC View**), the

Author: That's what this is called, I think?

<img width="218" alt="image" src="https://user-images.githubusercontent.com/1477535/175184107-9e4e89d7-f214-42a1-bb11-d244b55babaa.png">

> - In order to run experiments in the demo project (from the **DVC View**), the

Author: per https://code.visualstudio.com/docs/getstarted/userinterface#_views

> - In order to run experiments in the demo project (from the **DVC View**), the

Author: Called `dvc-views` under `views` in package.json which is slightly confusing (technically it's a single view with multiple panels for the side bar)

> - In order to run experiments in the demo project (from the **DVC View**), the

Author: We call it `dvc.viewContainer` (under `contributes.walkthroughs`) and file .../walkthrough/view-container.md but I think "container" is a term we came up with. I'm not seeing that in VS Code docs.

This is all irrelevant to the user of course but it may be good to decide on our own terminology. I vote for just "view" since it matches VS Code docs; I'm happy to update all these internals to singular `dvc(-view)` along with a walkthrough copy edit if needed.

> - In order to run experiments in the demo project (from the **DVC View**), the

Reviewer 2: How would you call then actual Views (Experiments, Sort By, Filter, etc) ?

Again, it's probably a bigger discussion. not relevant to this PR ... my comment was not about this at all

> - In order to run experiments in the demo project (from the **DVC View**), the

Author: Those are _Panels_ in the _Activity Bar_ of the _DVC View_.

Going back in the text in question, I guess you can run experiments both from several places (Experiments panel or table, or command palette) so no need to mention DVC View specifically... Linking from "run experiments" now instead. PTAL

> - In order to run experiments in the demo project (from the **DVC View**), the

Reviewer 2: I don't think they are panels. We can check the API or sources to see who it's called.

> - In order to run experiments in the demo project (from the **DVC View**), the

Reviewer: https://code.visualstudio.com/api/ux-guidelines/views#view-containers

> - In order to run experiments in the demo project (from the **DVC View**), the

Author: I see, thanks for clarifying VS Code UX terminology. And I see that `dvc-views` is also under `contributes .viewsContainers` indeed... So it's a _View Container_ for the _Primary Sidebar_.

Anyway, for now I just removed the reference and linked to the Run Experiments walkthroughs (PTAL).

But should we use official VS Code terminology (from the dev-facing docs) for README and walkthrough text going forward? I'd need to update "DVC View" in a few places now. Or do we want to have our own terminology like "click the **DVC** button in the _Activity Bar_" ?

> - In order to run experiments in the demo project (from the **DVC View**), the

Reviewer 2: > I guess there's some discrepancies

<img width="344" alt="Screen Shot 2022-06-24 at 5 55 26 PM" src="https://user-images.githubusercontent.com/3659196/175752007-0a379295-963d-475c-ace1-c74620bff14d.png">

no, there is no discrepancy I think. Most of the things (all?) in this list are Views. It just happened that in certain cases they are the only Views by default in the containers.

>   manually with the `dvc.dvcPath` setting (see [Configuration]).

Author: I'm using the MS Python ext but is the latter true? Can you set `dvc.dvcPath` to `.env/bin/dvc` even if `ms-python.python` is not configured to use `.env/bin/python`? ⏳

Author: > Can you set dvc.dvcPath to .env/bin/dvc even if ms-python.python is not configured to use .env/bin/python?

Tried that and it doesn't work (I get this issue: https://github.com/iterative/vscode-dvc/issues/1927). Changing this to require the `ms-python.python` for now... ✔️

  ```json
  "python.defaultInterpreterPath": "demo/.env/bin/dvc"
  ```

[dvc-checkpoints-mnist]:
  https://github.com/iterative/dvc-checkpoints-mnist/tree/make_checkpoint
[python extension]:
  https://marketplace.visualstudio.com/items?itemName=ms-python.python

<a id='warning'></a>

> **Warning** : When using any project that relies on an isolated Python
> environment (`conda`, `venv`, etc.), Microsoft's Python extension is required.

> > It's used by this extension to locate and utilize the required environment.

Author: Why is this the case? Why should some other extension ever be required for this extension to work?

Reviewer 2: this is fine, if you use Python it'll be installed. We need it since we are using its Python-specific APIs - like detecting and activating all these virtual envs

## The demo project

> The [demo project](demo) was initially forked from [dvc-checkpoints-mnist], and

Reviewer 2: let's remove this reference, it doesn't matter at all now I think

Author: Agreed. Esp. since https://github.com/iterative/dvc-checkpoints-mnist has no LICENSE (shoult it?)

Removed.

is provided as a lightweight, convenient testbed to try the extension out with.
It is not an exhaustive showcase of DVC's features, so testers are encouraged to
try other DVC repositories -- especially real-world cases!

- Install [Yarn](https://yarnpkg.com/) if needed. Run `yarn setup:venv` from the
  root of this repo to set up a virtual Python environment for the demo project
  (in `demo/.env`).

> - Go to the `demo/` directory of (`cd demo`) and run `yarn install` to set up

Reviewer 2: yes, it's not needed. `demo` itself even doesn't have the `package.json`.

>   the development environment.

Author: Last Q (I think): Is this really needed to play around with the demo project? Sounds more like a step for the previous section (**Developer environment**).

Author: UPDATE: Removed completely per https://github.com/iterative/vscode-dvc/pull/1928#pullrequestreview-1012834023

  > Stay in `demo/` if you want to use DVC from command line (terminal) in the
  > demo project.

- Pull the project data with the extension (**DVC panel** in the _Source Code_

>   view) or with `.env/bin/dvc pull` from command line.

Reviewer 2: active and run was better

Author: https://github.com/iterative/vscode-dvc/blob/contrib/demo/CONTRIBUTING.md#the-demo-project

  virtual env should be loaded through the Microsoft [Python extension] (please
  install it), having this in `.vscode/settings.json`:

  ```json

>   "python.defaultInterpreterPath": "demo/.env/bin/dvc"

Author: BTW why don't we just include this in the repo's .vscode/settings.json file, which is committed to Git?

Reviewer 2: this is not needed if you open `demo` directly

## The demo project

The [demo project](demo) is provided as a lightweight, convenient testbed to try
your changes to the extension.

> **Note**: It is not an exhaustive showcase of DVC's features. Testers are
> encouraged to try other DVC repositories -- especially real-world cases!

- Run `yarn setup:venv` from the root of this repo to set up a virtual Python
  environment for the demo project (in `demo/.env`).

> - Open the `demo/` folder of this repo directly in VS Code.

Reviewer 2: minor, I think "directly" is not needed here

Reviewer 2: open the folder - can be a bit ambiguous (open in the Editor vs Open as a Folder). But it looks like it's VS Code terminology. May be we can do "open the `demo/` project in VS Code"?

Author: Improved in 9f096503. PTAL

> - Pull the project data with the extension (either from the **DVC Tracked**

Reviewer 2: `with the extension` -> using the extension ...

Author: ```suggestion
- Pull the project data using the extension (either from the **DVC Tracked**
```

  panel in the _File Explorer_ or the **DVC panel** for _Source Control_) or by
  running `dvc pull` from a _Terminal_.

- In order to [run experiments] in the demo project, the virtual env should be

>   loaded through the Microsoft [Python extension] (please install it).

Reviewer 2: just [Python extension] ?

Reviewer 2: Python extension should be installed so that env can be activated ...

Author: ```suggestion
  loaded through the [Python extension] (please install it).
```

Author: Updated. PTAL @shcheklein

[python extension]:
  https://marketplace.visualstudio.com/items?itemName=ms-python.python
[run experiments]:
  https://github.com/iterative/vscode-dvc/blob/main/extension/resources/walkthrough/run-experiments.md

<a id='warning'></a>

> **Warning**: When using any project that relies on an isolated Python
> environment (`conda`, `venv`, etc.), Microsoft's Python extension is required.
> It's used by this extension to locate and utilize the required environment.

## React component development with Storybook

> Start Storybook with `yarn storybook` in either the monorepo root or in the

Reviewer 2: monorepo -> the repo root

Author: ```suggestion
Start Storybook with `yarn storybook` in either the root of this repo or in the
```

## The demo project

The [demo project](demo) is provided as a lightweight, convenient testbed to try
your changes to the extension.

> **Note**: It is not an exhaustive showcase of DVC's features. Testers are
> encouraged to try other DVC repositories -- especially real-world cases!

- Run `yarn setup:venv` from the root of this repo to set up a virtual Python
  environment for the demo project (in `demo/.env`).

- Open the `demo/` folder of this repo directly in VS Code.

- Pull the project data with the extension (either from the **DVC Tracked**
  panel in the _File Explorer_ or the **DVC panel** for _Source Control_) or by
  running `dvc pull` from a _Terminal_.

- In order to [run experiments] in the demo project, the virtual env should be
  loaded through the Microsoft [Python extension] (please install it), having
  this in `.vscode/settings.json`:

>   ```json

Reviewer 2: we probably want to remove this after all?

Author: Didn't get why, but it's been removed now.

Reviewer 2: sorry, to clarify. It should be detected automatically when you open the `demo/` project directly in a lot of cases. In some cases it will depend on how DVC is installed, etc. I feel we just can't cover properly different environments with this settings and in a lot of cases, if everything works find it's not needed

- Go to the `demo/` directory of this repo.

- Install [Yarn](https://yarnpkg.com/) if needed, and run `yarn install` to
  setting up the project.

- Go back to the root of this repo (not `./demo`) and `yarn setup:venv` to setup
  the Python virtual environment in the demo project.

>   > Python 3.8 is recommended, as some dependencies are not yet available for

Reviewer 2: hmm. I'm using Python 3.9 I think

Author: Same and so far no issues. This info was there already so I'm assuming it's correct.

Reviewer 2: got it , I think we can remove it

Author: Gone!

  > Python 3.9.

> - Return to `demo/`, activate the virtual environment, and run `dvc pull`.

Author: How do I activate the env? 😅 I'm not sure what kind of env it is even. There's no `activate` binary. I see:

```console
$ ls .env/bin 
__pycache__            dul-upload-pack        ftfy                   pip3                   python3.9              tqdm
cmark                  dulwich                futurize               pip3.10                shortuuid              wheel
convert-caffe2-to-onnx dvc                    jp.py                  pip3.9                 shtab
convert-onnx-to-caffe2 f2py                   normalizer             pygmentize             slugify
distro                 f2py3                  pasteurize             python                 tabulate
dul-receive-pack       f2py3.9                pip                    python3                torchrun
```

@iterative/vs-code 🙏🏼

Author: > There's no activate script.

I went through this again but running `yarn setup:env` as the first step, and there's an `activate` script now 🤷🏼 so this is answered I think. I'll update the instructions... ✔️

Author: Still not sure what kind of venv it is from the virtualenv-install.ts script, TBH. I guess readers don't need to know exactly but I'm curious: is it the same as a `python -m venv` ?

Reviewer 2: It is `python -m venv` (which is similar to `virtualenv`). It a bit suboptimal btw, since `wheel` is not installed - some packages can be slower / unavailable with this setup.

Author: OK. Should we make an issue to deprecate virtualenv-install.ts or make it into platform-specific shell scripts? TBH I think just mentioning "create a `virtualenv`" with some links in the README should be good enough.
